src/Microsoft.Azure.ServiceBus/ServiceBusDiagnosticsSource.cs

```diff
@@ -0,0 +1,771 @@
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region AddRule
+
+        internal Activity AddRuleStart(RuleDescription description)
+        {
+            return Start("AddRule", () => new
+            {
+                Rule = description,
+                Entity = this.entityPath,
+                Endpoint = this.endpoint
+            },
+            null);
+        }
+
+        internal void AddRuleStop(Activity activity, RuleDescription description, TaskStatus? status)
+        {
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    Rule = description,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    Status = status ?? TaskStatus.Faulted
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region RemoveRule
+
+        internal Activity RemoveRuleStart(string ruleName)
+        {
+            return Start("RemoveRule", () => new
+            {
+                RuleName = ruleName,
+                Entity = this.entityPath,
+                Endpoint = this.endpoint
+            },
+            null);
+        }
+
+        internal void RemoveRuleStop(Activity activity, string ruleName, TaskStatus? status)
+        {
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    RuleName = ruleName,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    Status = status ?? TaskStatus.Faulted
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region GetRules
+
+        internal Activity GetRulesStart()
+        {
+            return Start("GetRules", () => new
+            {
+                Entity = this.entityPath,
+                Endpoint = this.endpoint
+            },
+            null);
+        }
+
+        internal void GetRulesStop(Activity activity, IEnumerable<RuleDescription> rules, TaskStatus? status)
+        {
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    Rules = rules,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    Status = status ?? TaskStatus.Faulted
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region AcceptMessageSession
+
+        internal Activity AcceptMessageSessionStart(string sessionId)
+        {
+            return Start("AcceptMessageSession", () => new
+                {
+                    SessionId = sessionId,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint
+                },
+                a => a.AddTag("SessionId", sessionId));
+        }
```

Comment:
How good are these with null params? Couple of methods here have good chance of being passed a null param (like this one). What happens in that case?

Reply:
Nothing would break if there is a null value of sessionId, but I'd still prefer not to add null to Tags (just to optimize amount of data). 
I'll add null check then, thanks